### PR TITLE
chore: switch to global redis instance

### LINKED ISSUE #2519: Use global redis instance in core service
We now have a global redis instance configured at the renku chart level. The core service should either just use that global instance or make it configurable wether it will deploy its own redis. 
See https://github.com/SwissDataScienceCenter/renku/pull/2332 and https://github.com/SwissDataScienceCenter/renku-gateway/pull/488 for context.

### PR DESCRIPTION
**Pull request by Panaetius**
close #2519 

Shamelessly stolen from https://github.com/SwissDataScienceCenter/renku-gateway/pull/537

/deploy #persist

### CODE REVIEW

File: helm-chart/renku-core/templates/_helpers.tpl [hunk 1]
```diff
@@ -53,3 +53,16 @@ Hack for calling templates in a fake scope (until this is solved https://github.
...
+*/}}
+{{- define "redis.host" -}}
+{{- if .Values.global.redis.host -}}
+# If global hostname for redis is found use that
+{{- .Values.global.redis.host -}}
+{{- else -}}
+# If global hostname for redis is not found, then check for redis subchart and then use that
+{{- required "Either global.redis.host should be defined or a subchart for redis should be used where redis.fullname is available." (include "call-nested" (list . "redis" "redis.fullname")) -}}
+{{- end -}}
+{{- end -}}
```
Comment (olevski) [inline helm-chart/renku-core/templates/_helpers.tpl:68]:
```suggestion
{{- end -}}

```

Comment (olevski) [reply helm-chart/renku-core/templates/_helpers.tpl:68]: Just missing a new line at the end of the file.

File: helm-chart/renku-core/values.yaml [hunk 2]
```diff
@@ -78,23 +78,23 @@ resources: {}
 
 # affinity: {}
 
-## The renku core service deploys its own redis as a dependency. These values
-## do probably not need to be changed.
-redis:
-  nameOverride: core-redis
+## The renku core service uses the global redis instance of the renku platform.
```
Comment (olevski) [inline helm-chart/renku-core/values.yaml:81]: @Panaetius this whole section should be removed fully. The "hack" with `call-nested` essentially allows us to deploy this with renku. But you still need to specify all the other things in `.global.redis` - regardless of whether you have redis as a subchart or not.

If you want to be able to deploy with the redis instance defined here then we need to change the chart further.

**Review by olevski — CHANGES_REQUESTED** [review]

**Comment by olevski** [comment]
Also you should remove redis as a helm chart requirement from here: https://github.com/SwissDataScienceCenter/renku-python/blob/feature/2519-global-redis/helm-chart/renku-core/requirements.yaml#L2

**Comment by olevski** [comment]
One more thing. You need to add a `redis` section in the global portion of the values file. Like this:
```yaml
  ## Specify the information required to connect to a redis instance.
  redis:
    ## Host is optional, otherwise it needs to be provided below.
    ## The hostname is optional only if this chart is used as a child of the larger renku helm chart, in all other cases the value has to be provided below.
    # host: 
    ## The label used to enable the network policy that allows the gateway to connect to redis.
    ## If not defined below, it will default to <redis hostname>-client.
    # clientLabel:
    ## All the values below are required.
    isSentinel:
    dbIndex:
      core:
```

File: helm-chart/renku-core/values.yaml (hunk 2 above)
Comment (Panaetius) [reply helm-chart/renku-core/values.yaml:81]: I saw you kept this around in the gateway PR and thought we can do the same. but I don't think we want to deploy with an internal redis, so I'll remove it

File: helm-chart/renku-core/templates/deployment.yaml [hunk 3]
```diff
@@ -140,13 +143,16 @@ spec:
           args: ["worker"]
           env:
             - name: REDIS_HOST
-              value: {{ include "call-nested" (list $ "redis" "redis.fullname") }}-master
+              value: {{ template "redis.host" . }}
             - name: REDIS_PORT
               value: "6379"
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:148]:
```suggestion
              value: {{ .Values.global.redis.port | default "6379" | quote }}
```

Sorry I just noticed this

File: helm-chart/renku-core/templates/deployment.yaml [hunk 4]
```diff
@@ -185,13 +191,16 @@ spec:
...
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:196]:
```suggestion
              value: {{ .Values.global.redis.port | default "6379" | quote }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 5]
```diff
@@ -228,13 +237,16 @@ spec:
           args: ["scheduler"]
           env:
             - name: REDIS_HOST
-              value: {{ include "call-nested" (list $ "redis" "redis.fullname") }}-master
+              value: {{ template "redis.host" . }}
             - name: REDIS_PORT
               value: "6379"
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:242]:
```suggestion
              value: {{ .Values.global.redis.port | default "6379" | quote }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 6]
```diff
@@ -33,7 +33,7 @@ spec:
         app.kubernetes.io/name: {{ include "renku-core.name" $ }}
         app.kubernetes.io/instance: {{ $.Release.Name }}
         app.kubernetes.io/deploymentVersion: {{ $version.name }}
-        {{ include "call-nested" (list $ "redis" "redis.fullname") }}-client: "true"
+        {{ .Values.global.redis.clientLabel | default (printf "%s-%s" (include "redis.host" .) "-client") }}: "true"
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:36]:
```suggestion
        {{ $.Values.global.redis.clientLabel | default (printf "%s-%s" (include "redis.host" $) "-client") }}: "true"
```

Everything needs a `$` here. I missed that the whole deployment is in one `range` statement.

File: helm-chart/renku-core/templates/deployment.yaml [hunk 7]
```diff
@@ -185,13 +191,16 @@ spec:
           args: ["worker"]
           env:
             - name: REDIS_HOST
-              value: {{ include "call-nested" (list $ "redis" "redis.fullname") }}-master
+              value: {{ template "redis.host" . }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:194]:
```suggestion
              value: {{ template "redis.host" $ }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 8]
```diff
@@ -228,13 +237,16 @@ spec:
           args: ["scheduler"]
           env:
             - name: REDIS_HOST
-              value: {{ include "call-nested" (list $ "redis" "redis.fullname") }}-master
+              value: {{ template "redis.host" . }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:240]:
```suggestion
              value: {{ template "redis.host" $ }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 9]
```diff
@@ -185,13 +191,16 @@ spec:
           args: ["worker"]
           env:
             - name: REDIS_HOST
-              value: {{ include "call-nested" (list $ "redis" "redis.fullname") }}-master
+              value: {{ template "redis.host" . }}
             - name: REDIS_PORT
               value: "6379"
             - name: REDIS_DATABASE
-              value: "0"
+              value: {{ .Values.global.redis.dbIndex.coreService | quote }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:198]:
```suggestion
              value: {{ $.Values.global.redis.dbIndex.coreService | quote }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 10]
```diff
@@ -185,13 +191,16 @@ spec:
...
             - name: REDIS_PASSWORD
-              value:
+              valueFrom:
+                secretKeyRef:
+                  name: {{ .Values.global.redis.existingSecret }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:202]:
```suggestion
                  name: {{ $.Values.global.redis.existingSecret }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 11]
```diff
@@ -185,13 +191,16 @@ spec:
...
+                  key: {{ .Values.global.redis.existingSecretPasswordKey }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:203]:
```suggestion
                  key: {{ $.Values.global.redis.existingSecretPasswordKey }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 12]
```diff
@@ -228,13 +237,16 @@ spec:
...
             - name: REDIS_PORT
               value: "6379"
             - name: REDIS_DATABASE
-              value: "0"
+              value: {{ .Values.global.redis.dbIndex.coreService | quote }}
             - name: REDIS_PASSWORD
-              value:
+              valueFrom:
+                secretKeyRef:
+                  name: {{ .Values.global.redis.existingSecret }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:248]:
```suggestion
                  name: {{ $.Values.global.redis.existingSecret }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 13]
```diff
@@ -228,13 +237,16 @@ spec:
...
+                  key: {{ .Values.global.redis.existingSecretPasswordKey }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:249]:
```suggestion
                  key: {{ $.Values.global.redis.existingSecretPasswordKey }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 14]
```diff
@@ -228,13 +237,16 @@ spec:
           args: ["scheduler"]
           env:
             - name: REDIS_HOST
-              value: {{ include "call-nested" (list $ "redis" "redis.fullname") }}-master
+              value: {{ template "redis.host" . }}
             - name: REDIS_PORT
               value: "6379"
             - name: REDIS_DATABASE
-              value: "0"
+              value: {{ .Values.global.redis.dbIndex.coreService | quote }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:244]:
```suggestion
              value: {{ $.Values.global.redis.dbIndex.coreService | quote }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 15]
```diff
@@ -140,13 +143,16 @@ spec:
           args: ["worker"]
           env:
             - name: REDIS_HOST
-              value: {{ include "call-nested" (list $ "redis" "redis.fullname") }}-master
+              value: {{ template "redis.host" . }}
             - name: REDIS_PORT
               value: "6379"
             - name: REDIS_DATABASE
-              value: "0"
+              value: {{ .Values.global.redis.dbIndex.coreService | quote }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:150]:
```suggestion
              value: {{ $.Values.global.redis.dbIndex.coreService | quote }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 16]
```diff
@@ -140,13 +143,16 @@ spec:
...
             - name: REDIS_PASSWORD
-              value:
+              valueFrom:
+                secretKeyRef:
+                  name: {{ .Values.global.redis.existingSecret }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:154]:
```suggestion
                  name: {{ $.Values.global.redis.existingSecret }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 17]
```diff
@@ -140,13 +143,16 @@ spec:
...
+                  key: {{ .Values.global.redis.existingSecretPasswordKey }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:155]:
```suggestion
                  key: {{ $.Values.global.redis.existingSecretPasswordKey }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 18]
```diff
@@ -140,13 +143,16 @@ spec:
           args: ["worker"]
           env:
             - name: REDIS_HOST
-              value: {{ include "call-nested" (list $ "redis" "redis.fullname") }}-master
+              value: {{ template "redis.host" . }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:146]:
```suggestion
              value: {{ template "redis.host" $ }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 19]
```diff
@@ -52,7 +52,7 @@ spec:
           imagePullPolicy: {{ $.Values.metrics.image.pullPolicy }}
           env:
             - name: RQ_REDIS_HOST
-              value: {{ include "call-nested" (list $ "redis" "redis.fullname") }}-master
+              value: {{ template "redis.host" . }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:55]:
```suggestion
              value: {{ template "redis.host" $ }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 20]
```diff
@@ -72,13 +72,16 @@ spec:
           {{- end }}
           env:
             - name: REDIS_HOST
-              value: {{ include "call-nested" (list $ "redis" "redis.fullname") }}-master
+              value: {{ template "redis.host" . }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:75]:
```suggestion
              value: {{ template "redis.host" $ }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 21]
```diff
@@ -72,13 +72,16 @@ spec:
...
             - name: REDIS_PORT
-              value: "6379"
+              value: {{ .Values.global.redis.port | default "6379" | quote }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:77]:
```suggestion
              value: {{ $.Values.global.redis.port | default "6379" | quote }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 22]
```diff
@@ -72,13 +72,16 @@ spec:
...
             - name: REDIS_DATABASE
-              value: "0"
+              value: {{ .Values.global.redis.dbIndex.coreService | quote }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:79]:
```suggestion
              value: {{ $.Values.global.redis.dbIndex.coreService | quote }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 23]
```diff
@@ -72,13 +72,16 @@ spec:
...
             - name: REDIS_PASSWORD
-              value:
+              valueFrom:
+                secretKeyRef:
+                  name: {{ .Values.global.redis.existingSecret }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:83]:
```suggestion
                  name: {{ $.Values.global.redis.existingSecret }}
```

File: helm-chart/renku-core/templates/deployment.yaml [hunk 24]
```diff
@@ -72,13 +72,16 @@ spec:
...
+                  key: {{ .Values.global.redis.existingSecretPasswordKey }}
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:84]:
```suggestion
                  key: {{ $.Values.global.redis.existingSecretPasswordKey }}
```

**Review by olevski — CHANGES_REQUESTED** [review]
Sorry @Panaetius I missed the first time around that the whole deployment is in a "for loop" essentially. So you need to add the `$` to access the proper scope with all the values. I submitted suggestions for all the occurrences. Hopefully there is a "batch commit suggestions" functionality in git 😄

File: helm-chart/renku-core/templates/deployment.yaml [hunk 25]
```diff
@@ -33,7 +33,7 @@ spec:
         app.kubernetes.io/name: {{ include "renku-core.name" $ }}
         app.kubernetes.io/instance: {{ $.Release.Name }}
         app.kubernetes.io/deploymentVersion: {{ $version.name }}
-        {{ include "call-nested" (list $ "redis" "redis.fullname") }}-client: "true"
+        {{ $.Values.global.redis.clientLabel | default (printf "%s-%s" (include "redis.host" .) "-client") }}: "true"
```
Comment (olevski) [inline helm-chart/renku-core/templates/deployment.yaml:36]:
```suggestion
        {{ $.Values.global.redis.clientLabel | default (printf "%s-%s" (include "redis.host" $) "-client") }}: "true"
```

We missed this @Panaetius.

Comment (Panaetius) [reply helm-chart/renku-core/templates/deployment.yaml]: Thanks for the commit :)

**Comment by Panaetius** [comment]
You have the same admin permissions as everyone on the team. I think CI sometimes just doesn't start because its out of Runners (5 dependabot PRs running at the same time can do that)

**Comment by olevski** [comment]
@Panaetius as I mentioned I will re-work this PR to conform to what me and @ableuler discussed. And then all component charts can use the same or very similar ways of including the global redis in their charts.

**Comment by olevski** [comment]
I opened a new PR #2696 that replaces this one

**Pull request closed without merging**